src/main/java/org/jenkinsci/plugins/workflow/actions/ArgumentsAction.java

```diff
@@ -24,65 +24,137 @@
+    /** Used as a placeholder marker for {@link Step} arguments not stored for various reasons. */
     public enum NotStoredReason {
-        /** Denotes an unsafe value that cannot be stored/displayed due to sensitive info */
+        /** Denotes an unsafe value that cannot be stored/displayed due to sensitive info. */
         MASKED_VALUE,
 
-        /** Denotes an object that is too big to retain, such as strings exceeding {@link #MAX_STRING_LENGTH} */
+        /** Denotes an object that is too big to retain, such as strings exceeding {@link #MAX_RETAINED_LENGTH} */
         OVERSIZE_VALUE
     }
 
-    /** Largest string we'll persist in the Step, for performance reasons */
-    public static final int MAX_STRING_LENGTH = 1024;
+    /** Largest String, Collection, or array size we'll retain -- provides a rough size limit on any single field.
+     *  Set to 0 or -1 to remove length limits.
+     */
+    private static final int MAX_RETAINED_LENGTH = Integer.getInteger(ArgumentsAction.class.getName()+".maxRetainedLength", 1024);;
+
+    /**
+     * Provides a basic check if an object contains any excessively large single elements.
+     * This recursively checks for String length, Map size, and array/collection size.
+     * We're not trying to cap total size just identify problem elements.
+     * @param o Object to check, with null allowed since we may see null inputs
+     * @param maxElements Max number of elements for a collection/map or characters in a string, or <0 to ignore length rules.
+     * @return True if object (or one of the contained objects) exceeds maxElements size.
+     */
+    public static boolean isOverSized(@CheckForNull Object o, final int maxElements) {
+        if (maxElements <= 0 ) {
+            return false;
+        }
+        if (o == null || Primitives.isWrapperType(o.getClass()) || o.getClass().isEnum()) {
+            return false;
+        }
+        if (o instanceof CharSequence) {
+            return ((CharSequence) o).length() > maxElements;
+        }
+        if ((o instanceof Map || o instanceof Collection || o.getClass().isArray())) {
+            if (CollectionUtils.size(o) > maxElements) {
+                return true;
+            }
+            if (o instanceof Collection) {
+                for (Object element : (Collection)o) {
+                    if (isOverSized(element, maxElements)) {
+                        return true;
+                    }
+                }
+            }
+            if (o instanceof Object[]){
+                for (Object element : (Object[])o) {
+                    if (isOverSized(element, maxElements)) {
+                        return true;
+                    }
+                }
+            }
+            if (o instanceof Map) {
+                for(Map.Entry<?,?> entry : ((Map<?,?>)o).entrySet()) {
+                    if (isOverSized((Object)(entry.getKey()), maxElements) || isOverSized((Object)(entry.getValue()), maxElements)) {
```

Comment:
Generics complaints.   There's probably more than one way to solve it, but this does the trick.